In [4]:
# import

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [5]:
# Loading cleaned csv data

df = pd.read_csv('E:/Git/blore-house-data-with-gradient-boost-and-XGBoost/clean_data.csv')
df.isnull().sum()

area             0
location         0
bhk              0
bath             0
balcony          0
parking          0
furnishing       0
property_type    0
age              0
price            0
dtype: int64

### Model Creation

In [ ]:
# defining x and y

x = df.drop(columns='price')
y = np.log1p(df['price'])
x.shape

(981, 9)

In [ ]:
# Train Test split

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.30, random_state=42)

In [9]:
# model
from sklearn.ensemble import GradientBoostingRegressor

gbr = GradientBoostingRegressor(n_estimators=400,
    learning_rate=0.05,
    max_depth=2,
    subsample=0.9,
    random_state=42)
gbr.fit(x_train, y_train)
y_pred1 = gbr.predict(x_test)


In [10]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

print("MAE:", mean_absolute_error(y_test, y_pred1))
print("MSE:", mean_squared_error(y_test, y_pred1))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred1)))
print("R2 Score:", r2_score(y_test, y_pred1))


MAE: 0.23575812096180648
MSE: 0.08136900648615768
RMSE: 0.2852525310775659
R2 Score: 0.5474961138162557


In [11]:
# XGBoost Model
from xgboost import XGBRegressor
xgb = XGBRegressor(n_estimators=500,
    learning_rate=0.05,
    max_depth=3,
    subsample=0.8,
    colsample_bytree=0.8,
    reg_alpha=0.1,
    reg_lambda=1.0,
    random_state=42)
xgb.fit(x_train, y_train)
y_pred2 = xgb.predict(x_test)

In [ ]:
print("MAE:", mean_absolute_error(y_test, y_pred2))
print("MSE:", mean_squared_error(y_test, y_pred2))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred2)))
print("R2 Score:", r2_score(y_test, y_pred2))

MAE: 0.24000911221954385
MSE: 0.08746687426830879
RMSE: 0.29574799114839107
R2 Score: 0.5135850586366961


### Model Performance Conclusion

#### The Gradient Boosting Regressor achieved an R² score of ~0.55, indicating it explains around 55% of the variance in house prices on the test set. XGBoost with tuned hyperparameters achieved a slightly lower R² of ~0.51, suggesting that for this relatively small dataset, simpler boosting methods generalize better. Overall, the results provide a strong and honest baseline for price prediction, with further improvements possible through richer feature engineering and additional data.
